File: day1/code.php
<?php

class AdventOfCode
{
    public function part1(): int
    {
        $input = self::getInput();
        return $this->calcIncrementing($input);
    }

    public function part2()
    {
        $input = self::getInput();
        $reduce = array_reduce($input, static function ($acc, $item) use ($input): array {
            $offset = $acc[0];
            $list = array_slice($input, $offset, 3);
            return count($list) === 3 ? [++$offset, array_merge($acc[1], [array_sum($list)])] : [++$offset, $acc[1]];
        }, [0, []])[1];
        return $this->calcIncrementing($reduce);
    }

    private function calcIncrementing(array $input): ?int
    {
        return array_reduce($input, static function ($acc, $item): array {
            return $acc[1] !== null && (int)$item > (int)$acc[1] ? [++$acc[0], $item] : [$acc[0], $item];
        }, [0, null])[0];
    }

    private static function getInput(): array
    {
        return explode("\n", file_get_contents("input.txt"));
    }
}


$aoc = new AdventOfCode();
echo 'Part 1: '.$aoc->part1().PHP_EOL;
echo 'Part 2: '.$aoc->part2().PHP_EOL;

File: day1/code.ex
defmodule AdventOfCode do

  defp input do
    {:ok, data} = File.read("input.txt")
    data
    |> String.split("\n")
    |> Enum.map(fn i -> String.to_integer(i) end)
  end

  defp count_incrementing(list) when is_list(list) do
    list
    |> Enum.reduce({0, nil}, fn i, acc ->
      if elem(acc, 1) != nil and i > elem(acc, 1), do: {elem(acc, 0) + 1, i}, else: {elem(acc, 0), i}
    end)
    |> elem(0)
  end

  def run_part1() do
    input()
    |> count_incrementing()
  end

  def run_part2() do
    input = input()
    input
    |> Enum.reduce({0, []}, fn _i, acc ->
      offset = elem(acc, 0)
      case Enum.slice(input, offset, 3) do
        list when is_list(list) and length(list) == 3 -> {offset+1, [list | elem(acc, 1)]}
        _ -> {offset+1, elem(acc, 1)}
      end
    end)
    |> elem(1)
    |> Enum.reverse()
    |> Enum.map(fn item -> Enum.sum(item) end)
    |> count_incrementing()
  end

end

IO.puts("Part1: #{AdventOfCode.run_part1()}")
IO.puts("Part2: #{AdventOfCode.run_part2()}")
